In [9]:
import pandas as pd
import matplotlib.pyplot as plt

from collections import Counter

%matplotlib inline

## Let's load our dataset

The [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) method from pandas is our friend here, it keeps things nice and simple, it can be configurable for lot's of options, but the default settings work for us here. The one argument we specify is the URL where the file is hosted. You could visit this URL, download the CSV, and explore it in Excel if you want.

This reads the csv provided, and converts it into a [`pd.DataFrame`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) object.

In [4]:
df = pd.read_csv('https://storage.googleapis.com/pick-n-mix-1/convertcsv.csv')

print(df.info())

df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32844 entries, 0 to 32843
Data columns (total 6 columns):
 #   Column                                      Non-Null Count  Dtype 
---  ------                                      --------------  ----- 
 0   LSOA code (2011)                            32844 non-null  object
 1   LSOA name (2011)                            32844 non-null  object
 2   Local Authority District code (2019)        32844 non-null  object
 3   Local Authority District name (2019)        32844 non-null  object
 4   Index of Multiple Deprivation (IMD) Rank    32844 non-null  int64 
 5   Index of Multiple Deprivation (IMD) Decile  32844 non-null  int64 
dtypes: int64(2), object(4)
memory usage: 1.5+ MB
None


LSOA code (2011)           LSOA name (2011)  \
0        E01000001        City of London 001A   
1        E01000002        City of London 001B   
2        E01000003        City of London 001C   
3        E01000005        City of London 001E   
4        E01000006  Barking and Dagenham 016A   

  Local Authority District code (2019) Local Authority District name (2019)  \
0                            E09000001                       City of London   
1                            E09000001                       City of London   
2                            E09000001                       City of London   
3                            E09000001                       City of London   
4                            E09000002                 Barking and Dagenham   

   Index of Multiple Deprivation (IMD) Rank  \
0                                     29199   
1                                     30379   
2                                     14915   
3                                      8678   
4                                     14486   

   Index of Multiple Deprivation (IMD) Decile  
0                                           9  
1                                          10  
2                                           5  
3                                           3  
4                                           5

## Validation

Let's make sure that the data meets some of our expectations. The `Index of Multiple Deprivation (IMD) Rank` column is a ranking, therefore no two LSOA locations should have the same ranking. Let's check if that assumption holds true?

We can count the number of times each value occurs in a column with the [`.value_counts`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html) method. One important thing to note here, you might notice, when reading the documentation, this method is used from a [`pd.Series`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html) object, rather than a [`pd.DataFrame`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html) object.

Basically, each DataFrame is made up of Series objects. Each Series represents a column of data. A DataFrame is a collection of one or more columns.

In [12]:
df['Index of Multiple Deprivation (IMD) Rank'].value_counts()

17970    2
26064    2
6891     2
16434    2
25064    2
        ..
8881     1
15026    1
12979    1
2740     1
2049     1
Name: Index of Multiple Deprivation (IMD) Rank, Length: 32818, dtype: int64

We can see that some of these ranks occur more than once. We should check how often we're seeing ranks that occur more than once. Let's use a [`Counter`](https://docs.python.org/3.7/library/collections.html#collections.Counter) from the collections package, which comes with Python, to count the number of occurences of each rank.

In [11]:
Counter(df['Index of Multiple Deprivation (IMD) Rank'].value_counts())

Counter({2: 26, 1: 32792})

So what's this telling us?

This Counter instance shows that there are 26 ranks that occur twice, but 32792 ranks that occur once. So presumably two different LSOA are tied as the 17970th ranked IMD.

Our assumption was proved incorrect, but we were close, and now we have a better understanding of the dataset.

## Filtering

Filtering is an important tool when we're exploring and analysing data.

My preferred approach to this is using the [`Boolean indexing`](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#boolean-indexing) method.

Confusingly [`pd.filter`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.filter.html) subsets the data according to columns or rows specified, rather than by values.

Let's look for Manchester LSOAs.

In [34]:
df[(df['Local Authority District name (2019)'] == 'Manchester')]

LSOA code (2011) LSOA name (2011) Local Authority District code (2019)  \
4935         E01005061  Manchester 018A                            E08000003   
4936         E01005062  Manchester 018B                            E08000003   
4937         E01005063  Manchester 018C                            E08000003   
4938         E01005065  Manchester 018D                            E08000003   
4939         E01005066  Manchester 018E                            E08000003   
...                ...              ...                                  ...   
32759        E01033684  Manchester 013F                            E08000003   
32760        E01033685  Manchester 003F                            E08000003   
32761        E01033686  Manchester 028F                            E08000003   
32762        E01033687  Manchester 003G                            E08000003   
32763        E01033688  Manchester 013G                            E08000003   

      Local Authority District name (2019)  \
4935                            Manchester   
4936                            Manchester   
4937                            Manchester   
4938                            Manchester   
4939                            Manchester   
...                                    ...   
32759                           Manchester   
32760                           Manchester   
32761                           Manchester   
32762                           Manchester   
32763                           Manchester   

       Index of Multiple Deprivation (IMD) Rank  \
4935                                        724   
4936                                      11725   
4937                                       1101   
4938                                       1994   
4939                                       5152   
...                                         ...   
32759                                     16686   
32760                                     11316   
32761                                      5347   
32762                                      3067   
32763                                       262   

       Index of Multiple Deprivation (IMD) Decile  
4935                                            1  
4936                                            4  
4937                                            1  
4938                                            1  
4939                                            2  
...                                           ...  
32759                                           6  
32760                                           4  
32761                                           2  
32762                                           1  
32763                                           1  

[282 rows x 6 columns]

Let's go one step further, let's look at only LSOAs in Manchester, but also with an IMD decile greater than 8

In [32]:
df[(df['Local Authority District name (2019)'] == 'Manchester') &
   (df['Index of Multiple Deprivation (IMD) Decile'] >= 8)]

LSOA code (2011) LSOA name (2011) Local Authority District code (2019)  \
5017        E01005155  Manchester 037C                            E08000003   
5019        E01005157  Manchester 037D                            E08000003   
5031        E01005169  Manchester 045C                            E08000003   
5036        E01005174  Manchester 043C                            E08000003   
5137        E01005277  Manchester 038D                            E08000003   

     Local Authority District name (2019)  \
5017                           Manchester   
5019                           Manchester   
5031                           Manchester   
5036                           Manchester   
5137                           Manchester   

      Index of Multiple Deprivation (IMD) Rank  \
5017                                     25023   
5019                                     26894   
5031                                     24162   
5036                                     25466   
5137                                     25774   

      Index of Multiple Deprivation (IMD) Decile  
5017                                           8  
5019                                           9  
5031                                           8  
5036                                           8  
5137                                           8

## Simple Transformation